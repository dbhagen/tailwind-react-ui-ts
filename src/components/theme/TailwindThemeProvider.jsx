import React from 'react'
import PropTypes from 'prop-types'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

const TailwindThemeProvider = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

TailwindThemeProvider.propTypes = {
  theme: PropTypes.shape({}),
  children: PropTypes.node,
}

TailwindThemeProvider.defaultProps = {
  theme: {},
  children: undefined,
}

export default TailwindThemeProvider
