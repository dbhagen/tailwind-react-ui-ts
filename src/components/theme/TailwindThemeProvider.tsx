import React from 'react'
import merge from 'lodash.merge'

import defaultTheme from './defaultTheme'
import TailwindTheme from './TailwindTheme'

export type TailwindThemeProviderProps = {
  theme: {}
  children: React.ReactNode
}

const TailwindThemeProvider: React.FunctionComponent<TailwindThemeProviderProps> = ({ theme, children }) => {
  const mergedTheme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}

export default TailwindThemeProvider
