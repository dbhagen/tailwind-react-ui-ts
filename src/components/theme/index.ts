export { default as defaultTheme } from './defaultTheme'
export { default as TailwindTheme } from './TailwindTheme'
export { default as TailwindThemeProvider } from './TailwindThemeProvider'
export { default as withTheme } from './withTheme'
export { default as useThemeValue } from './useThemeValue'
