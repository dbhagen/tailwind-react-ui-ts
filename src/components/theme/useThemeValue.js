import { tailwindPropToClassName } from '../tailwind'

export default (prefix, value, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
