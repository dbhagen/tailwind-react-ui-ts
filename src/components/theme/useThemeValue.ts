import { tailwindPropToClassName } from '../tailwind'

export default (prefix: string, value: string, userClassNames = '') => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
