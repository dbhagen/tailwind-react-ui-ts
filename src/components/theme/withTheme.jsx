import React from 'react'

import TailwindTheme from './TailwindTheme'

const withTheme = Component => {
  const WithTheme = props => (
    <TailwindTheme.Consumer>
      {theme => <Component {...props} theme={theme} />}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}

export default withTheme
