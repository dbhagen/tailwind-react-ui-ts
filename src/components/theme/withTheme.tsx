import React from 'react'

import TailwindTheme from './TailwindTheme'

export type themeType = {
  prefix: string
  radius: string
  spacing: {
    zero: number
    sm: number
    md: number
    lg: number
    xl: number
  }
  container: {
    sm: string
    md: string
    lg: string
  }
  text: {
    size: {
      body: string[]
      title: string[]
    },
    family: {
      body: string
      subtitle: string
      title: string
    }
  }
  brandColors: {
    primary: string
    secondary: string
    success: string
    danger: string
    warning: string
    info: string
  }
  textColors: {
    body: string
    link: string
    linkDark: string
    emphasis: string
    on: {
      primary: string
      secondary: string
      success: string
      danger: string
      warning: string
      info: string
      dark: string
    }
  }
  surfaceColors: {
    default: string
    dark: string
    light: string
  }
  highlightOffset: number
  accentSize: number
  breakpoints: {
    sm: string
    md: string
    lg: string
    xl: string
  }
}

export type withThemeProps = {
  theme: themeType
}

const withTheme = <T extends withThemeProps>(WrappedComponent: React.ComponentType<T & withThemeProps>) => {
  const WithTheme = (props: T) =>
    <TailwindTheme.Consumer >
      {theme => < WrappedComponent {...props} theme={theme} />}
    </TailwindTheme.Consumer >


  WithTheme.displayName = `WithTheme(${WrappedComponent.displayName})`

  return WithTheme
}

export default withTheme
