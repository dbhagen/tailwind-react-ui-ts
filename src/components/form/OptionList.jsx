import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { List } from '../list'
import { withTheme } from '../theme'

class OptionList extends PureComponent {
  constructor(props) {
    super(props)

    this.state = {
      checked: [],
    }

    this.handleChange = this.handleChange.bind(this)
  }

  handleChange(e) {
    const { checkbox } = this.props
    const { checked } = this.state
    const { value } = e.target

    if (checkbox) {
      return this.setState({
        checked: checked.includes(value)
          ? checked.filter(checkedVal => checkedVal !== value)
          : [...checked, value],
      })
    }

    return this.setState({
      checked: [e.target.value],
    })
  }

  render() {
    const {
      theme,
      children,
      name,
      checkbox,
      field,
      invalid,
      ...rest
    } = this.props
    const { checked } = this.state
    const describedBy = [field.labelId, field.errorId, field.helpId].filter(
      by => by,
    )
    const isInvalid = field.invalid || invalid

    return (
      <List reset m={{ b: 0 }} {...rest}>
        {React.Children.map(children, child => {
          const value = child.props && child.props.value

          return React.cloneElement(child, {
            id: `${field.inputId}-${value}`,
            name: checkbox ? `${name}[]` : name,
            checked: checked.includes(value),
            onChange: this.handleChange,
            'aria-invalid': isInvalid || undefined,
            'aria-describedby': describedBy.length
              ? describedBy.join(' ')
              : undefined,
          })
        })}
      </List>
    )
  }
}

OptionList.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  children: PropTypes.node,
  name: PropTypes.string.isRequired,
  field: PropTypes.shape({
    inputId: PropTypes.string,
    invalid: PropTypes.bool,
    disabled: PropTypes.bool,
  }),
  checkbox: PropTypes.bool,
  invalid: PropTypes.bool,
}

OptionList.defaultProps = {
  children: undefined,
  checkbox: false,
  field: {},
  invalid: false,
}

export { OptionList as component }
export default withTheme(OptionList)
