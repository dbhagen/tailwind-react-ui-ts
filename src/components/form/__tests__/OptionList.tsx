import React from 'react'
import { shallow } from 'enzyme'

import { component as OptionList } from '../OptionList'
import { defaultTheme } from '../../theme'

import { Radio } from '..'

const setup = (testProps = {}) => {
  const props = Object.assign({ theme: defaultTheme, name: 'test' }, testProps)

  const wrapper = shallow(
    <OptionList {...props}>
      <Radio value="yes" label="Yes" />
      <Radio value="no" label="No" />
      <Radio value="maybe" label="Maybe" />
    </OptionList>,
  )

  return {
    props,
    wrapper,
  }
}

describe('OptionList', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup()

    expect(wrapper).toMatchSnapshot()
  })
})
