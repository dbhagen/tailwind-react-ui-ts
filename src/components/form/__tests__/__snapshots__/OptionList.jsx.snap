// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionList renders matching snapshot 1`] = `
<WithTheme(List)
  m={
    Object {
      "b": 0,
    }
  }
  reset={true}
>
  <Radio
    checked={false}
    id="undefined-yes"
    key=".0"
    label="Yes"
    name="test"
    onChange={[Function]}
    value="yes"
  />
  <Radio
    checked={false}
    id="undefined-no"
    key=".1"
    label="No"
    name="test"
    onChange={[Function]}
    value="no"
  />
  <Radio
    checked={false}
    id="undefined-maybe"
    key=".2"
    label="Maybe"
    name="test"
    onChange={[Function]}
    value="maybe"
  />
</WithTheme(List)>
`;
